File: tests/domain/decision/ReleaseFlowProgressionService.test.ts
import { afterAll, afterEach, beforeAll, describe, expect, it } from "vitest";
import { DataSource } from "typeorm";
import { ReleaseFlowProgressionService } from "../../../src/domain/decision/ReleaseFlowProgressionService";
import { ReleaseFlowService } from "../../../src/domain/releaseflow/ReleaseFlowService";
import { ReleaseFlowRepository } from "../../../src/domain/releaseflow/ReleaseFlowRepository";
import { RequestService } from "../../../src/domain/releaseflow/RequestService";
import { RequestRepository } from "../../../src/domain/releaseflow/RequestRepository";
import { TaskRepository } from "../../../src/domain/task/TaskRepository";
import {
  createAndInitTestDataSource,
  clearAllTables,
  seedReleaseFlow,
  seedRequest,
  seedTask,
} from "../../helpers/testDataSource";

describe("ReleaseFlowProgressionService", () => {
  let ds: DataSource;
  let flowRepo: ReleaseFlowRepository;
  let requestRepo: RequestRepository;
  let taskRepo: TaskRepository;
  let flowService: ReleaseFlowService;
  let requestService: RequestService;
  let progressionService: ReleaseFlowProgressionService;

  beforeAll(async () => {
    ds = await createAndInitTestDataSource();
    flowRepo = new ReleaseFlowRepository(ds);
    requestRepo = new RequestRepository(ds);
    taskRepo = new TaskRepository(ds);
    flowService = new ReleaseFlowService(ds, flowRepo, requestRepo);
    requestService = new RequestService(ds, requestRepo);
    progressionService = new ReleaseFlowProgressionService(
      taskRepo,
      requestRepo,
      requestService,
      flowService,
      flowRepo
    );
  });

  afterAll(async () => {
    await ds.destroy();
  });

  afterEach(async () => {
    await clearAllTables(ds);
  });

  describe("progressAfterDecision", () => {
    it("completes request when all tasks are terminal (Approved/Skipped)", async () => {
      const rf = await seedReleaseFlow(ds, { currentStage: "SIT" });
      const req = await seedRequest(ds, rf.id, { stage: "SIT" });
      const task1 = await seedTask(ds, req.id, { taskStatus: "Approved" });
      await seedTask(ds, req.id, { taskStatus: "Skipped" });

      await progressionService.progressAfterDecision(task1.id);

      const updatedRequest = await requestRepo.findById(req.id);
      expect(updatedRequest?.requestStatus).toBe("Completed");
    });

    it("completes flow when in PROD and request is completed", async () => {
      const rf = await seedReleaseFlow(ds, { currentStage: "PROD" });
      const req = await seedRequest(ds, rf.id, { stage: "PROD" });
      const task = await seedTask(ds, req.id, { taskStatus: "Approved" });

      await progressionService.progressAfterDecision(task.id);

      const updatedFlow = await flowRepo.findById(rf.id);
      expect(updatedFlow?.flowStatus).toBe("Completed");
    });

    it("advances flow to next stage when request completed in SIT", async () => {
      const rf = await seedReleaseFlow(ds, { currentStage: "SIT" });
      const req = await seedRequest(ds, rf.id, { stage: "SIT" });
      const task = await seedTask(ds, req.id, { taskStatus: "Approved" });

      await progressionService.progressAfterDecision(task.id);

      const updatedFlow = await flowRepo.findById(rf.id);
      expect(updatedFlow?.currentStage).toBe("UAT");
    });

    it("advances from UAT to PROD", async () => {
      const rf = await seedReleaseFlow(ds, { currentStage: "UAT" });
      const req = await seedRequest(ds, rf.id, { stage: "UAT" });
      const task = await seedTask(ds, req.id, { taskStatus: "Approved" });

      await progressionService.progressAfterDecision(task.id);

      const updatedFlow = await flowRepo.findById(rf.id);
      expect(updatedFlow?.currentStage).toBe("PROD");
    });

    it("auto-readies next Pending task when not all tasks are terminal", async () => {
      const rf = await seedReleaseFlow(ds, { currentStage: "SIT" });
      const req = await seedRequest(ds, rf.id, { stage: "SIT" });
      const task1 = await seedTask(ds, req.id, { taskStatus: "Approved", taskName: "task-1" });
      const task2 = await seedTask(ds, req.id, { taskStatus: "Pending", taskName: "task-2" });

      await progressionService.progressAfterDecision(task1.id);

      const updatedTask2 = await taskRepo.findById(task2.id);
      expect(updatedTask2?.taskStatus).toBe("Ready_For_Execution");
    });

    it("doesn't auto-ready if not Pending", async () => {
      const rf = await seedReleaseFlow(ds, { currentStage: "SIT" });
      const req = await seedRequest(ds, rf.id, { stage: "SIT" });
      const task1 = await seedTask(ds, req.id, { taskStatus: "Approved", taskName: "task-1" });
      const task2Original = await seedTask(ds, req.id, {
        taskStatus: "Ready_For_Execution",
        taskName: "task-2",
      });

      await progressionService.progressAfterDecision(task1.id);

      const tasks = await taskRepo.findByRequestId(req.id);
      const task2 = tasks.find((t) => t.id === task2Original.id);
      expect(task2?.taskStatus).toBe("Ready_For_Execution");
    });

    it("handles multiple Pending tasks – readies only the first", async () => {
      const rf = await seedReleaseFlow(ds, { currentStage: "SIT" });
      const req = await seedRequest(ds, rf.id, { stage: "SIT" });
      const task1 = await seedTask(ds, req.id, { taskStatus: "Approved", taskName: "task-1" });
      const task2 = await seedTask(ds, req.id, { taskStatus: "Pending", taskName: "task-2" });
      const task3 = await seedTask(ds, req.id, { taskStatus: "Pending", taskName: "task-3" });

      await progressionService.progressAfterDecision(task1.id);

      const updatedTask2 = await taskRepo.findById(task2.id);
      const updatedTask3 = await taskRepo.findById(task3.id);
      // First pending is readied
      expect(updatedTask2?.taskStatus).toBe("Ready_For_Execution");
      // Others remain Pending
      expect(updatedTask3?.taskStatus).toBe("Pending");
    });
  });
});

File: src/test/java/com/wwa/deploymentagent/domain/decision/ReleaseFlowProgressionServiceTest.java
package com.wwa.deploymentagent.domain.decision;

import com.wwa.deploymentagent.contracts.AgentId;
import com.wwa.deploymentagent.contracts.enums.*;
import com.wwa.deploymentagent.domain.releaseflow.ReleaseFlow;
import com.wwa.deploymentagent.domain.releaseflow.ReleaseFlowRepository;
import com.wwa.deploymentagent.domain.releaseflow.Request;
import com.wwa.deploymentagent.domain.releaseflow.RequestRepository;
import com.wwa.deploymentagent.domain.task.Task;
import com.wwa.deploymentagent.domain.task.TaskRepository;
import com.wwa.deploymentagent.helper.TestDataHelper;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest
@ActiveProfiles("test")
@Transactional
@DisplayName("ReleaseFlowProgressionService")
class ReleaseFlowProgressionServiceTest {

    @Autowired private ReleaseFlowProgressionService progressionService;
    @Autowired private TaskRepository taskRepository;
    @Autowired private RequestRepository requestRepository;
    @Autowired private ReleaseFlowRepository releaseFlowRepository;
    @Autowired private TestDataHelper helper;

    private ReleaseFlow releaseFlow;
    private Request request;

    @BeforeEach
    void setUp() {
        releaseFlow = helper.seedReleaseFlow();
        request = helper.seedRequest(releaseFlow, "SIT", RequestStatus.Pending, AgentId.DEPLOYMENT_AGENT);
    }

    @Test
    @DisplayName("marks request Completed when all tasks are terminal (Approved/Skipped)")
    void progressAfterDecision_allTasksTerminal_marksRequestCompleted() {
        Task t1 = helper.seedTask(request, TaskStatus.Approved);
        Task t2 = helper.seedTask(request, TaskStatus.Skipped);

        progressionService.progressAfterDecision(t1.getId());

        Request refreshed = requestRepository.findById(request.getId()).orElseThrow();
        assertThat(refreshed.getRequestStatus()).isEqualTo(RequestStatus.Completed);
    }

    @Test
    @DisplayName("advances to next pending task when not all tasks are terminal")
    void progressAfterDecision_pendingTaskExists_advancesToReady() {
        Task approved = helper.seedTask(request, TaskStatus.Approved);
        Task pending  = helper.seedTask(request, TaskStatus.Pending);

        progressionService.progressAfterDecision(approved.getId());

        Task refreshedPending = taskRepository.findById(pending.getId()).orElseThrow();
        assertThat(refreshedPending.getTaskStatus()).isEqualTo(TaskStatus.Ready_For_Execution);
    }

    @Test
    @DisplayName("does not advance the next task while a critical task is awaiting review")
    void progressAfterDecision_criticalAwaitingReview_blocksNextTask() {
        Task criticalReviewTask = helper.seedTask(request, TaskStatus.Awaiting_Review, true);
        Task pending = helper.seedTask(request, TaskStatus.Pending);

        progressionService.progressAfterDecision(criticalReviewTask.getId());

        Task refreshedPending = taskRepository.findById(pending.getId()).orElseThrow();
        assertThat(refreshedPending.getTaskStatus()).isEqualTo(TaskStatus.Pending);
    }

    @Test
    @DisplayName("advances Release Flow stage after SIT request is completed")
    void progressAfterDecision_sitCompleted_advancesToUat() {
        Task task = helper.seedTask(request, TaskStatus.Approved);

        progressionService.progressAfterDecision(task.getId());

        ReleaseFlow refreshed = releaseFlowRepository.findById(releaseFlow.getId()).orElseThrow();
        // With one task approved, SIT completes → flow should advance to UAT or Completed
        assertThat(refreshed.getCurrentStage()).isIn("UAT", "SIT");
        // Stage advances only if it was the sole task; confirm request is Completed
        Request refreshedReq = requestRepository.findById(request.getId()).orElseThrow();
        assertThat(refreshedReq.getRequestStatus()).isEqualTo(RequestStatus.Completed);
    }
}
